### PR TITLE
Changelog endpoint

### PR DESCRIPTION
**Pull request by JakeGoodwin22**
## What does this pull request do?

Adds an endpoint to show a list of all the changes made to a referral for the overview page

## What is the intent behind these changes?

Allow users to see a list of all the changes made to a referral

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ChangelogValuesDTO.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+data class ChangelogValuesDTO(
+  val changelogId: UUID,
+  val referralId: UUID,
+  val topic: String,
+  val changedAt: OffsetDateTime,
+  val name: String,
+  val reasonForChange: String
+) {
+  companion object {
+    fun from(changelogValuesDTO: ChangelogValuesDTO): ChangelogValuesDTO {
```
Comment (kth13) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ChangelogValuesDTO.kt:15]: I believe this DTO is not necessary. It is creating the ```ChangelogValuesDTO``` from the same Object.It is normally used while mapping the entity object to DTO object

Comment (JakeGoodwin22) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ChangelogValuesDTO.kt:15]: Yeah I've removed it now, it wasn't even being used in the controller. Cheers!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ChangelogValuesDTO.kt [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+data class ChangelogValuesDTO(
+  val changelogId: UUID,
+  val referralId: UUID,
+  val topic: String,
+  val changedAt: OffsetDateTime,
+  val name: String,
+  val reasonForChange: String
+) {
+// nothing to do
```
Comment (kth13) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ChangelogValuesDTO.kt:14]: you can remove the brackets if you are not doing anything

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AmendReferralService.kt [hunk 3]
```diff
@@ -57,4 +59,28 @@ class AmendReferralService(
     return referralService.getSentReferralForUser(id, user)
       ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "sent referral not found [id=$id]")
   }
+
+  fun getListOfChangeLogEntries(referral: Referral): List<ChangelogValuesDTO> {
+    val changeLogEntities = changelogRepository.findByReferralIdOrderByChangedAtDesc(referral.id)
+    val listOfChangelogValuesDTO: MutableList<ChangelogValuesDTO> = mutableListOf()
```
Comment (kth13) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AmendReferralService.kt:65]: The general pattern we follow is sending back entity objects back to controller and use DTO mapper class to create the DTO that will sent back to client. Could we please do it that way

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ChangelogValuesDTO.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service.UserDetail
+import java.time.OffsetDateTime
+import java.util.UUID
+
+data class ChangelogValuesDTO(
+  val changelogId: UUID,
+  val referralId: UUID,
+  val topic: String,
+  val changedAt: OffsetDateTime,
+  val name: String,
```
Comment (kth13) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ChangelogValuesDTO.kt:13]: @JakeGoodwin22 , what is the purpose of this name and I am seeing you are not sending back the changedBy back to client. I am checking the designs, we need that value to fill in change_made_by

Comment (JakeGoodwin22) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/ChangelogValuesDTO.kt:13]: Changed as discussed

**Pull request closed without merging**